DNS: More validation in Domain

Geod24 opened the pull request:
```
Properly validate the domain name when creating an instance.
There are now two ways to create a Domain:
- `fromSafeString`: Will validate its argument and `Domain` will be
  a wrapper around it.
- `fromString`: Does what the constructor was previously doing,
  plus some extra validation. Will always duplicate the input.

The constructor has been disabled has it was too error prone.
```

@mkykadir : I am not too happy about how this looks, so suggestions are welcome.
The main idea is that we should have a way to instantiate a Domain from a pre-validated string without allocating.

Geod24 committed:
DNS: More validation in Domain

Properly validate the domain name when creating an instance.
There are now two ways to create a Domain:
- `fromSafeString`: Will validate its argument and `Domain` will be
  a wrapper around it.
- `fromString`: Does what the constructor was previously doing,
  plus some extra validation. Will always duplicate the input.

The constructor has been disabled has it was too error prone.

Geod24 committed:
Trivial: Simplify SOA fromConfig method

mkykadir commented on source/agora/common/DNS.d:
```diff
@@ -720,11 +720,12 @@ public struct ResourceRecord
             deserializeFull!(TYPE)(&ctx.read, ctx.options),
             deserializeFull!(CLASS)(&ctx.read, ctx.options),
             deserializeFull!(uint)(&ctx.read, ctx.options),
+            T.RDATA(null),
```
Separated commit, maybe

Geod24 replied:
`@disable`ing the default ctor of `Domain` means that every struct that contains a `Domain` is not default-constructible anymore. But sure, will split it out in a prior commit.

mkykadir commented on source/agora/common/DNS.d:
```diff
@@ -911,13 +913,90 @@ public struct Domain
 
-    /// Creates domain name with always trailing root
-    public this (scope inout const(char)[] v) @trusted inout pure nothrow
+    @disable this () @safe pure nothrow @nogc;
+    @disable this (inout(char)[]) inout @safe pure nothrow @nogc;
```
Default `ctor` would be enough

Geod24 replied:
I will double check, but I think it wasn't, or that it created nonsensical errors.

Geod24 replied:
Now I remember. It will give you an error like this:
```
onlineapp.d(7): Error: constructor `onlineapp.Container.this()` is not callable using argument types `(typeof(null))`
onlineapp.d(7):        expected 0 argument(s), not 1
```

Instead of:
```
onlineapp.d(8): Error: constructor `onlineapp.Container.this` cannot be used because it is annotated with `@disable`
```

mkykadir commented on source/agora/common/DNS.d:
```diff
@@ -911,13 +913,90 @@ public struct Domain
+            assert(chunk.length < 64,
+                   "Domain instantiated with a label longer than 64 characters");
+            auto errIdx = Domain.checkLabel(chunk);
+            if (errIdx < chunk.length)
```
The way it is checked can look similar to version in `fromString`

mkykadir commented on source/agora/common/DNS.d:
```diff
@@ -911,13 +913,90 @@ public struct Domain
+            ensure(chunk.length < 64,
+                   "Domain instantiated with a label longer than 64 characters");
+            auto errIdx = Domain.checkLabel(chunk);
+            ensure(errIdx >= chunk.length,
```
Why `>=`? L#1149 checks with `==`

Geod24 replied:
It shouldn't make a diff, but will change.

mkykadir commented on source/agora/node/Registry.d:
```diff
@@ -851,7 +854,7 @@ private struct ZoneData
         foreach (const ref payload; this)
         {
-            reply.answers ~= payload.toRR(Domain(format("%s.%s",
+            reply.answers ~= payload.toRR(Domain.fromString(format("%s.%s",
```
This can also be `fromSafeString` for consistency with L#619.
PublicKey validates itself and `this.root` is already a `Domain`

Geod24 commented on source/agora/node/Registry.d:
```diff
@@ -479,8 +479,8 @@ unittest
     // We only need a valid mname
-    ZoneData zone;
-    zone.root = Domain("net.bosagora.io");
+    ZoneData zone = void; // FIXME
```
Oops

Geod24 replied:
Added a commit to fix this.

Geod24 committed:
DNS: Prepare for `@disable`d `Domain` ctor

Disabling a `struct` default constructor means that it disables
default construction for any struct that include an instance of this `struct`.
As we are going to disable `Domain`'s default ctor in the next commit,
we prepare this one by explicitly instantiating some `struct` / `union`
that include a `Domain` in them.

Geod24 committed:
DNS: Rename parsePublicKeyFromDomain to extractPublicKey

The function used to check that the key was in the domain,
which it why it required the context, but it doesn't do this
anymore nowadays, so we can just extract it in a standalone function.

mkykadir commented on source/agora/node/Registry.d:
```diff
@@ -472,45 +472,89 @@ public class NameRegistry: NameRegistryAPI
+        catch (Exception exc) return PublicKey.init;
+    }
+
+    // `findZone` ensures, `domain` is owned by us;
+    // thus this method should NEVER be `public`
```
This can be removed

mkykadir commented on source/agora/node/Registry.d:
```diff
@@ -472,60 +472,104 @@ public class NameRegistry: NameRegistryAPI
+        catch (Exception exc) return PublicKey.init;
+    }
+
+    // `findZone` ensures, `domain` is owned by us;
+    // thus this method should NEVER be `public`
```
These should be removed

Geod24 replied:
Good eye, fixed!

Geod24 commented:
Rebased

mkykadir commented on source/agora/node/Registry.d:
```diff
@@ -472,60 +472,102 @@ public class NameRegistry: NameRegistryAPI
+
+*******************************************************************************/
+
+private PublicKey extractPublicKey (in char[] domain) @safe
```
Regarding this change, since #2770 was using this as a function delegate in `TypedPayload.make`, that function can be changed to remove delegate parameter and directly calling this.

Geod24 replied:
Do we even call this function ? It would have stopped compiling if we did.

mkykadir replied:
Not right now but it will be called, I can do this in separate PR

Geod24 commented:
Merging as removing the unused function can be done in another commit.
